### PR TITLE
Add ListenerCollectionFactory and ListenerConfigurationChecker

### PR DESCRIPTION
**Pull request by viktorprogger**
This PR is about deleting of yiisoft/yii-event package. After it is merged and before `yii-event` package is removed these PR should be merged:

- yiisoft/yii-demo#206
- yiisoft/app#120
- yiisoft/app-api#33
- yiisoft/yii-console#97 (not necessary)

### CODE REVIEW

File: src/Support/InvalidEventConfigurationFormatException.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            EventConfigurator accepts an array in the following format:
+                [
+                    'eventName1' => [\$listener1, \$listener2, ...],
+                    'eventName2' => [\$listener3, \$listener4, ...],
+                    ...
+                ]
+        SOLUTION;
```
Comment (roxblnfk) [inline src/Support/InvalidEventConfigurationFormatException.php:26]:
```suggestion
        return <<<'SOLUTION'
            EventConfigurator accepts an array in the following format:
                [
                    'eventName1' => [$listener1, $listener2, ...],
                    'eventName2' => [$listener3, $listener4, ...],
                    ...
                ]
        SOLUTION;
```

File: src/Support/ListenerCollectionFactory.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                );
+            }
+
+            if (!is_array($listeners)) {
+                $type = gettype($listeners);
+
+                throw new InvalidEventConfigurationFormatException(
+                    "Event listeners for $eventName must be an array, $type given."
+                );
+            }
```
Comment (roxblnfk) [inline src/Support/ListenerCollectionFactory.php:46]: Can you use `ListenerConfigurationChecker` instead?

File: src/Support/InvalidListenerConfigurationException.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+class InvalidListenerConfigurationException extends InvalidArgumentException implements FriendlyExceptionInterface
+{
+    public function getName(): string
+    {
+        return 'Invalid event listener configuration';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
```
Comment (roxblnfk) [inline src/Support/InvalidListenerConfigurationException.php:19]:
```suggestion
        return <<<'SOLUTION'
```

File: composer.json [hunk 4]
```diff
@@ -20,7 +20,10 @@
     "prefer-stable": true,
     "require": {
         "php": "^7.4|^8.0",
-        "psr/event-dispatcher": "1.0.0"
+        "psr/container": "^1.0",
+        "psr/event-dispatcher": "1.0.0",
```
Comment (roxblnfk) [inline composer.json:24]:
```suggestion
        "psr/event-dispatcher": "^1.0",
```

File: src/Support/ListenerCollectionFactory.php (hunk 2 above)
Comment (viktorprogger) [reply src/Support/ListenerCollectionFactory.php:46]: I think these errors are different: they are about wrong configuration format, while `ListenerConfigurationChecker` checks if the given listeners are callables. With wrong configuration format we're not able to attach event listeners, while with wrong listeners in a right format we can attach, but can't call them. `ListenerConfigurationChecker` is for the second case.

File: src/Support/InvalidEventConfigurationFormatException.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        return 'Configuration format passed to EventConfigurator is invalid';
+    }
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            EventConfigurator accepts an array in the following format:
+                [
+                    'eventName1' => [\$listener1, \$listener2, ...],
```
Comment (yiiliveext) [inline src/Support/InvalidEventConfigurationFormatException.php:22]: eventClassName1 is more clear

File: src/Support/ListenerCollectionFactory.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    public function __construct(Injector $injector, ContainerInterface $container)
+    {
+        $this->injector = $injector;
+        $this->container = $container;
+    }
+
+    /**
+     * @param array $eventListeners Event listener list in format ['eventName1' => [$listener1, $listener2, ...]]
+     * @param bool $precheck If true, all the listeners will be checked strictly if they are valid callables
```
Comment (yiiliveext) [inline src/Support/ListenerCollectionFactory.php:25]: The param $prechek does not exist.

File: src/Support/ListenerConfigurationChecker.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    public function check(array $configuration): void
+    {
+        foreach ($configuration as $eventName => $listeners) {
+            if (!is_string($eventName)) {
```
Comment (yiiliveext) [inline src/Support/ListenerConfigurationChecker.php:22]:
```php
if (!(is_string($eventName) && class_exists($eventName)))
```

File: src/Support/ListenerCollectionFactory.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+     * @return ListenerCollection
+     */
+    public function create(array $eventListeners): ListenerCollection
+    {
+        $listenerCollection = new ListenerCollection();
+
+        foreach ($eventListeners as $eventName => $listeners) {
+            if (!is_string($eventName) || !class_exists($eventName)) {
+                throw new InvalidEventConfigurationFormatException(
+                    'Incorrect event listener format. Format with event name must be used.'
```
Comment (yiiliveext) [inline src/Support/ListenerCollectionFactory.php:38]:
```suggestion
                    "Incorrect event format.  '$eventName' must be event class name."
```

File: src/Support/ListenerConfigurationChecker.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+     * @param array $configuration An array in format of [eventClassName => [listeners]]
+     *
+     * @psalm-suppress InvalidCatch
+     */
+    public function check(array $configuration): void
+    {
+        foreach ($configuration as $eventName => $listeners) {
+            if (!is_string($eventName)) {
+                throw new InvalidEventConfigurationFormatException(
+                    'Incorrect event listener format. Format with event name must be used.'
```
Comment (yiiliveext) [inline src/Support/ListenerConfigurationChecker.php:38]:
```suggestion
                    "Incorrect event format.  '$eventName' must be event class name."
```

File: src/Support/ListenerCollectionFactory.php (hunk 8 above)
Comment (samdark) [reply src/Support/ListenerCollectionFactory.php:38]: Let's not use single quotes in the end-user text. Double quotes only.

File: src/Support/ListenerConfigurationChecker.php (hunk 9 above)
Comment (samdark) [reply src/Support/ListenerConfigurationChecker.php:38]: Let's not use single quotes in the end-user text. Double quotes only.

**Comment by samdark** [comment]
Decided that these changes should end up in yii-event: https://gist.github.com/samdark/80de533cf160e5df99d570dfe4d414b3

**Pull request closed without merging**